File: aidlink/src/main/java/com/ibroadlink/library/aidlink/AidlinkBinderImpl.java
package com.ibroadlink.library.aidlink;

import android.os.Binder;
import android.os.RemoteCallbackList;
import android.os.RemoteException;

/**
 * Created by codezjx on 2017/9/13.<br/>
 */
final class AidlinkBinderImpl extends ITransfer.Stub implements AidlinkBinder {
    
    private static final String TAG = "LinkerBinder";
    private RemoteCallbackList<ICallback> mCallbackList;
    private Invoker mInvoker;

    AidlinkBinderImpl() {
        mInvoker = new Invoker();
        mCallbackList = mInvoker.getCallbackList();
    }

    @Override
    public void registerObject(Object target) {
        mInvoker.registerObject(target);
    }

    @Override
    public void unRegisterObject(Object target) {
        mInvoker.unRegisterObject(target);
    }
    
    @Override
    public Response execute(Request request) throws RemoteException {
        for (BaseTypeWrapper wrapper : request.getArgsWrapper()) {
            Logger.d(TAG, "Receive param, value:" + wrapper.getParam()
                    + " type:" + (wrapper.getParam() != null ? wrapper.getParam().getClass() : "null"));
        }
        Logger.d(TAG, "Receive request:" + request.getMethodName());
        return mInvoker.invoke(request);
    }

    @Override
    public void register(ICallback callback) throws RemoteException {
        int pid = Binder.getCallingPid();
        Logger.d(TAG, "register callback:" + callback + " pid:" + pid);
        if (callback != null) {
            mCallbackList.register(callback, pid);
        }
    }

    @Override
    public void unRegister(ICallback callback) throws RemoteException {
        int pid = Binder.getCallingPid();
        Logger.d(TAG, "unRegister callback:" + callback + " pid:" + pid);
        if (callback != null) {
            mCallbackList.unregister(callback);
        }
    }
}

File: aidlink/src/main/java/com/ibroadlink/library/aidlink/AidlinkBinderImpl.kt
package com.ibroadlink.library.aidlink

import android.os.RemoteCallbackList
import android.os.RemoteException

/**
 * Created by codezjx on 2017/9/13.<br></br>
 */
internal class AidlinkBinderImpl : ITransfer.Stub(), AidlinkBinder {
    private val mCallbackList: RemoteCallbackList<ICallback>
    private val mInvoker: Invoker = Invoker()
    override fun registerObject(target: Any?) {
        mInvoker.registerObject(target)
    }

    override fun unRegisterObject(target: Any?) {
        mInvoker.unRegisterObject(target)
    }

    @Throws(RemoteException::class)
    override fun execute(request: Request): Response {
        for (wrapper in request.argsWrapper) {
            Logger.d(
                TAG, "Receive param, value:" + wrapper.param
                        + " type:" + if (wrapper.param != null) wrapper.param.javaClass else "null"
            )
        }
        Logger.d(TAG, "Receive request:" + request.methodName)
        return mInvoker.invoke(request)
    }

    @Throws(RemoteException::class)
    override fun register(callback: ICallback) {
        val pid = getCallingPid()
        Logger.d(TAG, "register callback:$callback pid:$pid")
        mCallbackList.register(callback, pid)
    }

    @Throws(RemoteException::class)
    override fun unRegister(callback: ICallback) {
        val pid = getCallingPid()
        Logger.d(TAG, "unRegister callback:$callback pid:$pid")
        mCallbackList.unregister(callback)
    }

    companion object {
        private const val TAG = "LinkerBinder"
    }

    init {
        mCallbackList = mInvoker.callbackList
    }
}